src/ocean/Assets.ts

```diff
@@ -239,82 +239,96 @@ export class Assets extends Instantiable {
-          oldDdo.service[i].attributes.additionalInformation = Object()
+    if (!ddo) ddo = await this.ocean.metadatacache.retrieveDDO(did)
+    if (!ddo) return null
+    for (let i = 0; i < ddo.service.length; i++) {
+      if (ddo.service[i].type === 'metadata') {
+        if (newMetadata.title) ddo.service[i].attributes.main.name = newMetadata.title
+        if (!ddo.service[i].attributes.additionalInformation)
+          ddo.service[i].attributes.additionalInformation = Object()
         if (newMetadata.description)
-          oldDdo.service[i].attributes.additionalInformation.description =
+          ddo.service[i].attributes.additionalInformation.description =
             newMetadata.description
         if (newMetadata.links)
-          oldDdo.service[i].attributes.additionalInformation.links = newMetadata.links
-      }
-    }
-    if (newMetadata.servicePrices) {
-      for (i = 0; i < newMetadata.servicePrices.length; i++) {
-        if (
-          newMetadata.servicePrices[i].cost &&
-          newMetadata.servicePrices[i].serviceIndex
-        ) {
-          oldDdo.service[newMetadata.servicePrices[i].serviceIndex].attributes.main.cost =
-            newMetadata.servicePrices[i].cost
-        }
+          ddo.service[i].attributes.additionalInformation.links = newMetadata.links
       }
     }
-    const storeTx = await this.ocean.OnChainMetadataCache.update(
-      oldDdo.id,
-      oldDdo,
-      account.getId()
-    )
-    if (storeTx) return oldDdo
-    else return null
+    return ddo
   }
 
   /**
    * Update Compute Privacy
-   * @param  {did} string DID.
-   * @param  {number} serviceIndex Index of the compute service in the DDO
+   * @param  {did} string DID. You can leave this empty if you already have the DDO
+   * @param  {ddo} DDO if empty, will trigger a retrieve
+   * @param  {number} serviceIndex Index of the compute service in the DDO. If -1, will try to find it
    * @param  {ServiceComputePrivacy} computePrivacy ComputePrivacy fields & new values.
    * @param  {Account} account Ethereum account of owner to sign and prove the ownership.
-   * @return {Promise<string>}
+   * @return {Promise<DDO>}
    */
   public async updateComputePrivacy(
     did: string,
+    ddo: DDO,
     serviceIndex: number,
-    computePrivacy: ServiceComputePrivacy,
-    account: Account
+    computePrivacy: ServiceComputePrivacy
   ): Promise<DDO> {
-    const oldDdo = await this.ocean.metadatacache.retrieveDDO(did)
-    if (oldDdo.service[serviceIndex].type !== 'compute') return null
-    oldDdo.service[serviceIndex].attributes.main.privacy.allowRawAlgorithm =
+    if (!ddo) ddo = await this.ocean.metadatacache.retrieveDDO(did)
+    if (!ddo) return null
+    if (serviceIndex === -1) {
+      const service = ddo.findServiceByType('compute')
+      if (!service) return null
+      serviceIndex = service.index
+    }
+    if (typeof ddo.service[serviceIndex] === 'undefined') return null
```

Comment:
all these null returns could be done in one if statement in line 293 instead of sprinkling the conditions throughout multiple lines:

```
if (
  !ddo ||
  ddo.service[serviceIndex].type !== 'compute' ||
  typeof ddo.service[serviceIndex] === 'undefined'
)
  return null
```

Reply:
actually, the logic is:
 - try to get the ddo if we only have the DID (if we have the DDO already, avoid extra call to aqua)
 - if serviceIndex == -1 -> try to search for compute service
 - else check if serviceIndex is a valid compute service (maybe it's not)